modules/stochastic_tools/src/reporters/ParallelSolutionStorage.C

```diff
@@ -0,0 +1,144 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ParallelSolutionStorage.h"
+#include "NonlinearSystemBase.h"
+
+registerMooseObject("StochasticToolsApp", ParallelSolutionStorage);
+
+InputParameters
+ParallelSolutionStorage::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Parallel container to store serialized solution fields from "
+                             "simulations on sub-applications.");
+  params.addRequiredParam<std::vector<VariableName>>(
+      "variables",
+      "The names of the variables whose serialized solution this object is supposed to receive.");
+
+  // Making this required to make sure nobody prints full solution vectors on accident
+  params.makeParamRequired<std::vector<OutputName>>("outputs");
+  return params;
+}
+
+ParallelSolutionStorage::ParallelSolutionStorage(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _distributed_solutions(
+        declareValueByName<
+            std::map<VariableName,
+                     std::unordered_map<unsigned int, std::vector<DenseVector<Real>>>>>(
+            "parallel_solution_storage", REPORTER_MODE_DISTRIBUTED)),
+    _variable_names(getParam<std::vector<VariableName>>("variables"))
+{
+  for (const auto & vname : _variable_names)
+  {
+    _distributed_solutions.emplace(
+        vname, std::unordered_map<unsigned int, std::vector<DenseVector<Real>>>());
+  }
+}
+
+void
+ParallelSolutionStorage::initialSetup()
+{
+  for (const auto & vname : _variable_names)
+    _distributed_solutions[vname].clear();
+}
+
+void
+ParallelSolutionStorage::addEntry(const VariableName & vname,
+                                  unsigned int global_i,
+                                  const DenseVector<Real> & solution)
+{
+  mooseAssert(std::find(_variable_names.begin(), _variable_names.end(), vname) !=
+                  _variable_names.end(),
+              "We are trying to add a variable that we cannot receive!");
+
+  auto sample_insert_pair =
+      _distributed_solutions[vname].emplace(global_i, std::vector<DenseVector<Real>>());
+
+  sample_insert_pair.first->second.push_back(std::move(solution));
+}
+
+unsigned int
+ParallelSolutionStorage::totalNumberOfStoredSolutions(const VariableName & vname) const
+{
+  const auto & samples = libmesh_map_find(_distributed_solutions, vname);
+
+  unsigned int count = 0;
+  for (const auto & sample : samples)
+    count += sample.second.size();
```

Comment:
Thanks for the suggestion but I don't necessarily see how accumulate is nicer than this. I would need to write a lambda which accumulates the sizes of the map.second-s which is equally long.

Reply:
Read https://stackoverflow.com/questions/3188793/when-should-the-stl-algorithms-be-used-instead-of-using-your-own

They express more intent to both the developer and the compiler

Reply:
it's good training for when it's really needed. Codewise here the for-loop is easier to maintain.

Reply:
Okay, changed it!

Reply:
C++17 makes it super easy to turn on threaded parallelization of STL algorithms, too, and NVidia's made some progress on GPGPU parallelization of them.  I also wouldn't blink twice at this for loop here, but at some point in the not too distant future I anticipate going over a hundred thousand lines of code turning loops into `std::` calls and feeling regretful that we didn't get into the habit of using `<algorithm>` more from the start.

Reply:
Ooh, and even when we're full to the brim on MPI and don't have threads to spare, marking an algorithm as parallelizable might speed things up.  A loop that unrolls to `x = x0 + x1 + x2 + x3 + x4 + x5 + x6 +x7` isn't supposed to be turned into `x = (x0 + x1 + x2 + x3) + (x4 + x5 + x6 + x7)`, because that changes roundoff error and makes IEEE sad, and you don't want to enable associativity in general via compiler flags because a few linear algebra algorithms really depend on that operations order, but if you tell the compiler that it can pretend FP addition is associative in a particular case then it's allowed to use SIMD instructions to evaluate the second form much more quickly.